### PR TITLE
feat(idempotency): add custom JMESPath functions

### PR DESCRIPTION
**Pull request by dreamorosi**
## Description of your changes

This PR introduces support for custom JMESPath functions to the Idempotency utility, addressing one of our most requested features ([the 5th](https://github.com/aws-powertools/powertools-lambda-typescript/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc) to be precise).

With this new feature customers can use the `powertools_json()`, `powertools_base64()`, and `powertools_base64_gzip()` custom functions to deserialize encoded payloads and use the values within them as idempotency key.

This is especially useful when working with API Gateway or SQS messages among others, since oftentimes the customer defined part of the payload is encoded as JSON stringified object or base64 encoded value:

```ts
import { makeIdempotent, IdempotencyConfig } from '@aws-lambda-powertools/idempotency';
import { DynamoDBPersistenceLayer } from '@aws-lambda-powertools/idempotency/dynamodb';
import type { Context, APIGatewayProxyEvent } from 'aws-lambda';

const persistenceStore = new DynamoDBPersistenceLayer({
  tableName: 'idempotencyTableName',
});

const myHandler = async (
  event: APIGatewayProxyEvent,
  _context: Context
): Promise<void> => {
  // your code goes here here
};

export const handler = makeIdempotent(myHandler, {
  persistenceStore,
  config: new IdempotencyConfig({
    eventKeyJmespath: 'powertools_json(body).["user", "productId"]',
  }),
});
```

The feature is added by having the Idempotency utility take a dependency on the newly released `@aws-lambda-powertools/jmespath` package and replacing the existing, and more limited, `jmespath` module.

The PR also adds the new `@aws-lambda-powertools/jmespath` utility to the Lambda Layer replacing the existing `jmespath` dependency.

### Related issues, RFCs

**Issue number:** #1298

## Checklist

- [x] [My changes meet the tenets criteria](https://docs.powertools.aws.dev/lambda/typescript/#tenets)
- [x] I have performed a *self-review* of my own code
- [x] I have *commented* my code where necessary, particularly in areas that should be flagged with a TODO, or hard-to-understand areas
- [x] I have made corresponding changes to the *documentation*
- [x] My changes generate *no new warnings*
- [x] I have *added tests* that prove my change is effective and works
- [x] The PR title follows the [conventional commit semantics](https://github.com/aws-powertools/powertools-lambda-typescript/blob/main/.github/semantic.yml#L2)

### Breaking change checklist

***Is it a breaking change?:*** NO

- [ ] I have documented the migration process
- [ ] I have added, implemented necessary warnings (if it can live side by side)

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

### CODE REVIEW

**Comment by dreamorosi** [comment]
Converting back to draft to fix an error appearing in the examples app.

**Comment by dreamorosi** [comment]
Integration tests passing: https://github.com/aws-powertools/powertools-lambda-typescript/actions/runs/8701862065

File: packages/idempotency/src/IdempotencyHandler.ts [hunk 1]
```diff
@@ -275,8 +276,9 @@ export class IdempotencyHandler<Func extends AnyFunction> {
       !this.#idempotencyConfig.throwOnNoIdempotencyKey
     ) {
       const selection = search(
+        this.#idempotencyConfig.eventKeyJmesPath,
         this.#functionPayloadToBeHashed,
-        this.#idempotencyConfig.eventKeyJmesPath
+        { customFunctions: new PowertoolsFunctions() }
```
Comment (am29d) [inline packages/idempotency/src/IdempotencyHandler.ts:281]: Is there any reason to create new object every time? This pattern goes doen to `IdempotencyHandler`. I am wondering if we can use this as a private field, this object is the same everywhere, we can use it once.

**Review by am29d — COMMENTED** [review]
Look good, just minor comment on instantiation of `PowertoolsFunction()`

**Review by am29d — APPROVED** [review]

**Pull request merged**